apollo-router-core/src/services/http_compat.rs

```diff
@@ -47,23 +115,33 @@ impl<T> Request<T> {
     /// This is the constructor (or builder) to use when constructing a "fake" Request.
     ///
     /// This does not enforce the provision of the uri and method that is required for a fully functional
-    /// Request. It's usually enough for testing, when a fully consructed Request is
+    /// Request. It's usually enough for testing, when a fully constructed Request is
     /// difficult to construct and not required for the purposes of the test.
+    ///
+    /// In addition, fake requests are expected to be valid, and will panic if given invalid values.
     pub fn fake_new(
-        headers: Vec<(HeaderName, HeaderValue)>,
+        headers: MultiMap<IntoHeaderName, IntoHeaderValue>,
         uri: Option<http::Uri>,
         method: Option<http::Method>,
         body: T,
-    ) -> http::Result<Request<T>> {
+    ) -> Request<T> {
```

Comment:
I don't think we should be varying the return type for the fake and the real constructor. That will lead to problems in testing at some point.

Reply:
Happy to go the other way. We had a mix before though.

Reply:
In case we are talking at cross purposes...

What we have (i.e. before this PR) is a `new` and a `fake_new` function which both return `http::Result<Request<T>> `. I don't mind what the return is, but it should be the same for both. That's the invariant that's important to maintain (I think) with these fake/real constructors, so there is no extraneous variation between real and test behaviour.

Does your comment mean there are other constructors not maintaining this invariant? That's probably a problem if so.

Reply:
@garypen  I made the change in [68d0a33](https://github.com/apollographql/router/pull/908/commits/68d0a33f2fd5ec6769ba4f0d3ff81f356af8e168)

This may be worth double checking. It added a ton of noise to the code, and not sure of the benefit for the fakes as they are for use in testing only and add a ton of noise. 

I can see both sides of the argument though, and it does dissuade people from using the fakes inside real code just to avoid the having to deal with `Result`. Take a look at the commit and it still seems like the right direction then I'm happy to go with it,

Reply:
We decided to go with consistency.